File: code/packages/haskell/reed-solomon/src/ReedSolomon.hs
-- | ReedSolomon — Reed-Solomon error-correcting codes over GF(256).
--
-- Reed-Solomon is a block error-correcting code: given a message of @k@
-- bytes, the encoder adds @nCheck@ redundancy bytes such that a decoder
-- can recover the original message even if up to @t = nCheck \/ 2@ of the
-- @k + nCheck@ bytes are corrupted.
--
-- == Where RS is Used
--
-- * __QR codes__ — up to 30% of a QR symbol can be damaged and still decode.
-- * __CDs\/DVDs__ — two-level CIRC corrects scratches.
-- * __Hard drives__ — sector-level error correction firmware.
-- * __Voyager probes__ — images transmitted across billions of kilometres.
-- * __RAID-6__ — two parity drives are exactly an @(n, n-2)@ RS code.
--
-- == Building Blocks
--
-- @
--   MA00  polynomial   — GF(256) polynomial arithmetic
--   MA01  gf256        — GF(2^8) field arithmetic (add=XOR, mul=table)
--   MA02  reed-solomon — THIS MODULE
-- @
--
-- == Systematic Encoding
--
-- The output is __systematic__: the original message bytes appear unchanged
-- at the front of the codeword, followed by the computed check bytes:
--
-- @
--   output = [ message[0] ... message[k-1] | check[0] ... check[nCheck-1] ]
-- @
--
-- == Polynomial Convention
--
-- Codeword bytes are treated as a __big-endian__ polynomial:
--
-- @
--   codeword[0]·x^{n-1} + codeword[1]·x^{n-2} + ... + codeword[n-1]
-- @
--
-- Syndrome evaluation uses this big-endian (highest-degree-first) convention.
-- The generator polynomial and Berlekamp-Massey state use little-endian
-- (index = degree) internally.
module ReedSolomon
  ( -- * Errors
    RSError (..)

    -- * Encoding
  , buildGenerator
  , encode

    -- * Decoding
  , decode
  , syndromes
  , errorLocator
  ) where

import Data.List (foldl')
import GF256     (GF256, gfAdd, gfSub, gfMul, gfDiv, gfPow, gfInv, gfZero, gfOne)

-- ---------------------------------------------------------------------------
-- Error Types
-- ---------------------------------------------------------------------------

-- | Errors that can occur during RS encoding or decoding.
data RSError
    = TooManyErrors
      -- ^ The codeword has more than @t = nCheck\/2@ corrupted bytes.
      --   The data is unrecoverable.
    | InvalidInput String
      -- ^ The input parameters are invalid (e.g. @nCheck@ is odd or 0,
      --   or the total codeword length exceeds 255).
    deriving (Show, Eq)

-- ---------------------------------------------------------------------------
-- Internal Polynomial Helpers
-- ---------------------------------------------------------------------------

-- | Evaluate a __big-endian__ GF(256) polynomial at @x@.
--
-- @p[0]@ is the highest-degree coefficient. Horner's method left-to-right:
--
-- @
--   acc = 0
--   for each byte b in p (high degree first):
--     acc = acc·x XOR b   (all in GF(256))
-- @
polyEvalBE :: [GF256] -> GF256 -> GF256
polyEvalBE p x = foldl' step gfZero p
  where
    step acc b = gfAdd (gfMul acc x) b

-- | Evaluate a __little-endian__ GF(256) polynomial at @x@ (Horner).
--
-- @p[i]@ is the coefficient of @xⁱ@. Iterates from high to low degree.
polyEvalLE :: [GF256] -> GF256 -> GF256
polyEvalLE p x = foldl' step gfZero (reverse p)
  where
    step acc c = gfAdd (gfMul acc x) c

-- | Multiply two __little-endian__ GF(256) polynomials (convolution).
--
-- @result[i+j] ^= a[i] · b[j]@  (XOR-accumulate).
polyMulLE :: [GF256] -> [GF256] -> [GF256]
polyMulLE [] _ = []
polyMulLE _ [] = []
polyMulLE as bs = foldr combine (replicate resultLen gfZero) indexed
  where
    resultLen = length as + length bs - 1
    indexed   = [ (i + j, gfMul a b)
                | (i, a) <- zip [0..] as
                , (j, b) <- zip [0..] bs ]
    combine (k, v) xs = take k xs ++ [gfAdd (xs !! k) v] ++ drop (k + 1) xs

-- | Compute the remainder of __big-endian__ polynomial division in GF(256).
--
-- The divisor must be __monic__ (leading coefficient = 1).
--
-- Algorithm: for each step, eliminate the current leading term by
-- subtracting a scaled copy of the divisor:
--
-- @
--   for i = 0 .. (len(dividend) - len(divisor)):
--     coeff = dividend[i]
--     for j = 0 .. len(divisor):
--       dividend[i+j] ^= coeff · divisor[j]
--
--   remainder = last (divLen - 1) elements
-- @
polyModBE :: [GF256] -> [GF256] -> [GF256]
polyModBE dividend divisor
    | length dividend < divLen = dividend
    | otherwise                = drop (length rem' - (divLen - 1)) rem'
  where
    divLen  = length divisor
    steps   = length dividend - divLen + 1
    rem'    = foldl' elimStep (dividend ++ []) [0..steps - 1]

    elimStep :: [GF256] -> Int -> [GF256]
    elimStep buf i =
        let coeff = buf !! i
        in  if coeff == gfZero
              then buf
              else
                let updates = zip [i..] (map (gfMul coeff) divisor)
                    applyUpdate xs (k, v) = take k xs ++ [gfAdd (xs !! k) v] ++ drop (k + 1) xs
                in  foldl' applyUpdate buf updates

-- | Compute the inverse error locator for byte position @p@ in a codeword
-- of length @n@.
--
-- In big-endian convention, position @p@ has degree @n-1-p@.
-- The locator is @X_p = α^{n-1-p}@, so @X_p⁻¹ = α^{(p+256-n) mod 255}@.
invLocator :: Int -> Int -> GF256
invLocator p n =
    let e = (p + 256 - n) `mod` 255
    in  gfPow 2 e

-- ---------------------------------------------------------------------------
-- Generator Polynomial
-- ---------------------------------------------------------------------------

-- | Build the Reed-Solomon generator polynomial for @nCheck@ check bytes.
--
-- The generator is the product of @nCheck@ linear factors:
--
-- @
--   g(x) = (x + α¹)(x + α²)…(x + α^{nCheck})
-- @
--
-- where @α = 2@ is the primitive element of GF(256).
--
-- Returns a __little-endian__ coefficient array (index = degree) of length
-- @nCheck + 1@. The last element is always @1@ (monic polynomial).
--
-- Algorithm:
--
-- @
--   g = [1]
--   for i = 1 to nCheck:
--     new_g[j] = GF256.mul(αⁱ, g[j]) XOR g[j-1]
-- @
--
-- Example: @nCheck = 2@
--
-- @
--   Start: g = [1]
--   i=1: α¹ = 2;  g = [gfMul 2 1, 1] = [2, 1]
--   i=2: α² = 4;  new[0] = gfMul 4 2 = 8
--                 new[1] = gfMul 4 1 XOR 2 = 4 XOR 2 = 6
--                 new[2] = 1
--                 g = [8, 6, 1]
-- @
--
-- Returns 'Left' ('InvalidInput') if @nCheck@ is 0 or odd.
buildGenerator :: Int -> Either RSError [GF256]
buildGenerator nCheck
    | nCheck == 0 || nCheck `mod` 2 /= 0 =
        Left (InvalidInput ("nCheck must be a positive even number, got " ++ show nCheck))
    | otherwise =
        Right (foldl' multiplyFactor [gfOne] [1..nCheck])
  where
    multiplyFactor :: [GF256] -> Int -> [GF256]
    multiplyFactor g i =
        let alphaI = gfPow 2 i
            -- New polynomial has length (length g + 1).
            -- new[j] = gfMul alphaI g[j] XOR g[j-1]
            -- where g[-1] = 0 and g[length g] = 0.
            newLen  = length g + 1
            newG    = [ let curr = if j < length g then gfMul alphaI (g !! j) else gfZero
                            prev = if j > 0        then g !! (j - 1)          else gfZero
                        in  gfAdd curr prev
                      | j <- [0..newLen - 1]
                      ]
        in  newG

-- ---------------------------------------------------------------------------
-- Encoding
-- ---------------------------------------------------------------------------

-- | Encode a message with Reed-Solomon, producing a systematic codeword.
--
-- __Systematic__ means the message bytes are unchanged in the output:
--
-- @
--   output = [ message bytes | check bytes ]
-- @
--
-- Algorithm:
--
-- 1. Build generator @g@ (little-endian of length @nCheck+1@), reverse to
--    big-endian @g_BE@.
-- 2. Form @shifted = message ++ replicate nCheck 0@ (represents
--    @M(x)·x^{nCheck}@ in big-endian).
-- 3. @remainder = polyModBE shifted g_BE@
-- 4. Output @message ++ remainder@ (padded to exactly @nCheck@ bytes).
--
-- Why it works: @C(x) = M(x)·x^{nCheck} + R(x) = Q(x)·g(x)@, so
-- @C(αⁱ) = 0@ for @i = 1…nCheck@ — all roots of @g@ are roots of @C@.
--
-- Returns 'Left' on invalid input.
encode :: [GF256] -> Int -> Either RSError [GF256]
encode message nCheck
    | nCheck == 0 || nCheck `mod` 2 /= 0 =
        Left (InvalidInput ("nCheck must be a positive even number, got " ++ show nCheck))
    | n > 255 =
        Left (InvalidInput ("total codeword length " ++ show n ++ " exceeds GF(256) block size limit of 255"))
    | otherwise =
        case buildGenerator nCheck of
          Left err -> Left err
          Right gLE ->
            let gBE     = reverse gLE          -- big-endian divisor
                shifted = message ++ replicate nCheck gfZero
                rmd     = polyModBE shifted gBE
                -- Pad remainder to exactly nCheck bytes
                padded  = replicate (nCheck - length rmd) gfZero ++ rmd
            in  Right (message ++ padded)
  where
    n = length message + nCheck

-- ---------------------------------------------------------------------------
-- Decoding
-- ---------------------------------------------------------------------------

-- | Compute the @nCheck@ syndromes of a received codeword.
--
-- @S_j = received(α^j)@ for @j = 1, …, nCheck@.
--
-- If all syndromes are zero, the codeword has no errors.
-- The codeword is evaluated as a __big-endian__ polynomial.
syndromes :: [GF256] -> Int -> [GF256]
syndromes received nCheck =
    [ polyEvalBE received (gfPow 2 j) | j <- [1..nCheck] ]

-- | Berlekamp-Massey algorithm: find the error locator polynomial.
--
-- Given syndromes @S₁, …, S_{2t}@, find the shortest LFSR generating
-- the syndrome sequence. Returns @(Λ, L)@ where @Λ@ is the error locator
-- polynomial (little-endian, @Λ[0] = 1@) and @L@ is the number of errors.
--
-- Algorithm:
--
-- @
--   C = [1], B = [1], L = 0, xShift = 1, b = 1
--
--   for n = 0 to 2t-1:
--     d = S[n] XOR ∑_{j=1}^{L} C[j]·S[n-j]
--
--     if d == 0:
--       xShift++
--     elif 2L <= n:
--       T = C; C = C XOR (d\/b)·x^{xShift}·B; L = n+1-L; B = T; b = d; xShift = 1
--     else:
--       C = C XOR (d\/b)·x^{xShift}·B; xShift++
-- @
berlekampMassey :: [GF256] -> ([GF256], Int)
berlekampMassey synds = go 0 [gfOne] [gfOne] 0 1 gfOne
  where
    twoT = length synds

    go :: Int -> [GF256] -> [GF256] -> Int -> Int -> GF256 -> ([GF256], Int)
    go n c b bigL xShift bScale
        | n >= twoT = (c, bigL)
        | otherwise =
            let -- Discrepancy: d = S[n] + sum_{j=1..L} C[j] * S[n-j]
                d = foldl' (\acc j ->
                              if j < length c && n >= j
                                then gfAdd acc (gfMul (c !! j) (synds !! (n - j)))
                                else acc)
                            (synds !! n)
                            [1..bigL]

            in  if d == gfZero
                  then go (n + 1) c b bigL (xShift + 1) bScale

                  else if 2 * bigL <= n
                    then
                      let tSave  = c
                          scale  = gfDiv d bScale
                          -- c = c XOR scale * x^xShift * b
                          c'     = xorShiftScaled c xShift scale b
                          bigL'  = n + 1 - bigL
                      in  go (n + 1) c' tSave bigL' 1 d

                    else
                      let scale = gfDiv d bScale
                          c'    = xorShiftScaled c xShift scale b
                      in  go (n + 1) c' b bigL (xShift + 1) bScale

    -- XOR (scale * x^shift * poly) into acc.
    xorShiftScaled :: [GF256] -> Int -> GF256 -> [GF256] -> [GF256]
    xorShiftScaled acc shift scale poly =
        let -- Shifted poly: [0, 0, ..., 0, scale*poly[0], scale*poly[1], ...]
            scaledShifted = replicate shift gfZero ++ map (gfMul scale) poly
            -- XOR into acc, extending if needed.
            len   = max (length acc) (length scaledShifted)
            pad xs = xs ++ replicate (len - length xs) gfZero
        in  zipWith gfAdd (pad acc) (pad scaledShifted)

-- | Exposed error locator computation (Berlekamp-Massey).
--
-- Returns @Λ(x)@ in __little-endian__ form with @Λ[0] = 1@.
--
-- Useful for diagnostics and QR decoder components.
errorLocator :: [GF256] -> [GF256]
errorLocator = fst . berlekampMassey

-- | Chien Search: find which byte positions contain errors.
--
-- Position @p@ is an error location if @Λ(X_p⁻¹) = 0@, where
-- @X_p⁻¹ = α^{(p+256-n) mod 255}@ for a codeword of length @n@.
chienSearch :: [GF256] -> Int -> [Int]
chienSearch lambda n =
    [ p | p <- [0..n-1], polyEvalLE lambda (invLocator p n) == gfZero ]

-- | Forney Algorithm: compute error magnitudes from error positions.
--
-- For each error at position @p@:
--
-- @
--   e_p = Ω(X_p⁻¹) \/ Λ'(X_p⁻¹)
-- @
--
-- where:
--
-- * @Ω(x) = (S(x) · Λ(x)) mod x^{2t}@ — error evaluator polynomial
-- * @S(x) = S₁ + S₂x + … + S_{2t}x^{2t-1}@ — syndrome polynomial (LE)
-- * @Λ'(x)@ — formal derivative of @Λ@ in characteristic 2
--
-- In GF(2), the formal derivative keeps only __odd-indexed__ coefficients:
--
-- @
--   Λ'(x) = Λ₁ + Λ₃x² + Λ₅x⁴ + …
-- @
--
-- Returns 'Left' 'TooManyErrors' if any denominator evaluates to zero.
forney :: [GF256] -> [GF256] -> [Int] -> Int -> Either RSError [GF256]
forney lambda synds positions n =
    let twoT = length synds

        -- Ω = S(x) · Λ(x) mod x^{2t}: truncate to first 2t terms.
        omegaFull = polyMulLE synds lambda
        omega     = take twoT omegaFull

        -- Formal derivative of Λ in characteristic 2.
        --
        -- For Λ = [λ₀, λ₁, λ₂, λ₃, λ₄, ...]:
        --   Λ'(x) = λ₁ + 0·x + λ₃·x² + 0·x³ + λ₅·x⁴ + …
        --
        -- Rule: Λ'[k] = Λ[k+1] when (k+1) is odd, else 0.
        --   = Λ[k+1] when k is even.
        --
        -- Implementation: iterate over even k values, reading Λ[k+1].
        lp = computeLambdaPrime lambda

    in  mapM (magnitude omega lp) positions
  where
    -- Build the formal derivative polynomial.
    -- For each even k in [0, 2, 4, ...], Λ'[k] = Λ[k+1].
    -- Odd positions are 0 (coefficient of x^{odd} in characteristic 2).
    computeLambdaPrime :: [GF256] -> [GF256]
    computeLambdaPrime lam
        | null lam  = []
        | otherwise =
            let maxIdx = length lam - 2  -- last valid index for lam[k+1]
                -- Build list of length (length lam - 1): Λ'[k] for k=0..len-2
                result = [ if even k then lam !! (k + 1) else gfZero
                         | k <- [0..maxIdx] ]
            in  result

    magnitude omega lp pos =
        let xiInv    = invLocator pos n
            omegaVal = polyEvalLE omega xiInv
            lpVal    = polyEvalLE lp   xiInv
        in  if lpVal == gfZero
              then Left TooManyErrors
              else Right (gfDiv omegaVal lpVal)

-- | Decode a received Reed-Solomon codeword, correcting up to @t = nCheck\/2@
-- errors.
--
-- Pipeline:
--
-- @
--   received
--     │
--     ▼  Step 1: Syndromes S₁…S_{nCheck}
--     │          all zero? → return message directly
--     │
--     ▼  Step 2: Berlekamp-Massey → Λ(x), error count L
--     │          L > t? → TooManyErrors
--     │
--     ▼  Step 3: Chien search → error positions {p₁…pᵥ}
--     │          |positions| ≠ L? → TooManyErrors
--     │
--     ▼  Step 4: Forney → error magnitudes {e₁…eᵥ}
--     │
--     ▼  Step 5: Correct received[pₖ] XOR= eₖ
--     │
--     ▼  Return first k = len - nCheck bytes
-- @
--
-- Returns 'Left' 'TooManyErrors' if the codeword has more than @t@ errors,
-- or 'Left' 'InvalidInput' if parameters are invalid.
decode :: [GF256] -> Int -> Either RSError [GF256]
decode received nCheck
    | nCheck == 0 || nCheck `mod` 2 /= 0 =
        Left (InvalidInput ("nCheck must be a positive even number, got " ++ show nCheck))
    | length received < nCheck =
        Left (InvalidInput ("received length " ++ show (length received)
                             ++ " < nCheck " ++ show nCheck))
    | otherwise =
        let t    = nCheck `div` 2
            nn   = length received
            k    = nn - nCheck

            -- Step 1: Syndromes
            synds = syndromes received nCheck

        in  if all (== gfZero) synds
              then Right (take k received)
              else
                let -- Step 2: Berlekamp-Massey
                    (lambda, numErrors) = berlekampMassey synds

                in  if numErrors > t
                      then Left TooManyErrors
                      else
                        let -- Step 3: Chien Search
                            positions = chienSearch lambda nn

                        in  if length positions /= numErrors
                              then Left TooManyErrors
                              else
                                -- Step 4: Forney
                                case forney lambda synds positions nn of
                                  Left err         -> Left err
                                  Right magnitudes ->
                                    -- Step 5: Apply corrections
                                    let corrected = applyCorrections received (zip positions magnitudes)
                                    in  Right (take k corrected)
  where
    -- Apply XOR corrections at error positions.
    applyCorrections :: [GF256] -> [(Int, GF256)] -> [GF256]
    applyCorrections bytes [] = bytes
    applyCorrections bytes ((pos, mag):rest) =
        let updated = take pos bytes ++ [gfAdd (bytes !! pos) mag] ++ drop (pos + 1) bytes
        in  applyCorrections updated rest

File: code/packages/c/reed-solomon/src/reed_solomon.c
/*
 * reed_solomon.c — implementation of Reed-Solomon coding (see reed_solomon.h).
 * A faithful port of the Rust `reed-solomon` crate: generator polynomial,
 * systematic encoding, and the syndromes -> Berlekamp-Massey -> Chien -> Forney
 * decode pipeline, all over GF(2^8) via the sibling gf256 package.
 *
 * Every polynomial here is bounded by the GF(256) block size (255 bytes), so all
 * working storage lives in fixed stack buffers — no heap allocation.
 */
#include "reed_solomon.h"

#include <string.h> /* memcpy, memset */

#include "gf256.h" /* gf256_add / multiply / divide / power (default 0x11D) */

#define RS_BUF 512 /* comfortably above any codeword/product length (<= ~382) */

/* ── GF(256) polynomial helpers ───────────────────────────────────────────── */
/* Evaluate a big-endian polynomial (p[0] = highest degree) at x (Horner). */
static uint8_t poly_eval_be(const uint8_t *p, size_t plen, uint8_t x) {
    uint8_t acc = 0;
    size_t i;
    for (i = 0; i < plen; i++) {
        acc = gf256_add(gf256_multiply(acc, x), p[i]);
    }
    return acc;
}
/* Evaluate a little-endian polynomial (p[i] = coeff of x^i) at x. */
static uint8_t poly_eval_le(const uint8_t *p, size_t plen, uint8_t x) {
    uint8_t acc = 0;
    size_t i;
    for (i = plen; i > 0; i--) {
        acc = gf256_add(gf256_multiply(acc, x), p[i - 1]);
    }
    return acc;
}
/* Multiply two little-endian polynomials (convolution) into out; returns len. */
static size_t poly_mul_le(const uint8_t *a, size_t alen, const uint8_t *b,
                          size_t blen, uint8_t *out) {
    size_t outlen, i, j;
    if (alen == 0 || blen == 0) {
        return 0;
    }
    outlen = alen + blen - 1;
    memset(out, 0, outlen);
    for (i = 0; i < alen; i++) {
        for (j = 0; j < blen; j++) {
            out[i + j] = gf256_add(out[i + j], gf256_multiply(a[i], b[j]));
        }
    }
    return outlen;
}
/* Remainder of big-endian division by a monic divisor; returns len (divlen-1). */
static size_t poly_mod_be(const uint8_t *dividend, size_t dlen,
                          const uint8_t *divisor, size_t divlen, uint8_t *out) {
    uint8_t rem[RS_BUF];
    size_t steps, i, j;
    memcpy(rem, dividend, dlen);
    if (dlen < divlen) {
        memcpy(out, rem, dlen);
        return dlen;
    }
    steps = dlen - divlen + 1;
    for (i = 0; i < steps; i++) {
        uint8_t coeff = rem[i];
        if (coeff == 0) {
            continue;
        }
        for (j = 0; j < divlen; j++) {
            rem[i + j] = gf256_add(rem[i + j], gf256_multiply(coeff, divisor[j]));
        }
    }
    memcpy(out, rem + (dlen - (divlen - 1)), divlen - 1);
    return divlen - 1;
}

/* ── generator polynomial ─────────────────────────────────────────────────── */
/* n_check must be even and >= 2, and small enough that a codeword still fits in
 * a GF(256) block (255 bytes) — with at least a 1-byte message, n_check <= 254.
 * The upper bound is also what keeps every fixed working buffer in bounds. */
static int valid_n_check(size_t n_check) {
    return n_check != 0 && n_check % 2 == 0 && n_check <= 254;
}

rs_status rs_build_generator(size_t n_check, uint8_t *out, size_t *out_len) {
    uint8_t g[RS_BUF];
    size_t glen = 1;
    size_t i, j;
    if (!valid_n_check(n_check)) {
        return RS_INVALID_INPUT;
    }
    g[0] = 1;
    for (i = 1; i <= n_check; i++) {
        uint8_t alpha_i = gf256_power(2, (uint32_t)i);
        uint8_t new_g[RS_BUF];
        memset(new_g, 0, glen + 1);
        for (j = 0; j < glen; j++) {
            new_g[j] = gf256_add(new_g[j], gf256_multiply(g[j], alpha_i));
            new_g[j + 1] = gf256_add(new_g[j + 1], g[j]);
        }
        memcpy(g, new_g, glen + 1);
        glen++;
    }
    memcpy(out, g, glen);
    *out_len = glen;
    return RS_OK;
}

/* ── encoding ─────────────────────────────────────────────────────────────── */
rs_status rs_encode(const uint8_t *message, size_t message_len, size_t n_check,
                    uint8_t *out, size_t *out_len) {
    uint8_t g_le[RS_BUF], g_be[RS_BUF], shifted[RS_BUF], rem[RS_BUF];
    size_t glen, remlen, n, i;
    if (!valid_n_check(n_check)) {
        return RS_INVALID_INPUT;
    }
    n = message_len + n_check;
    if (n > 255 || n < message_len) { /* second test guards size_t overflow */
        return RS_INVALID_INPUT;
    }
    rs_build_generator(n_check, g_le, &glen); /* glen = n_check + 1 */
    for (i = 0; i < glen; i++) {
        g_be[i] = g_le[glen - 1 - i]; /* big-endian, monic (g_be[0] = 1) */
    }
    /* shifted = message || n_check zeros  (= M(x) * x^n_check). */
    if (message_len) {
        memcpy(shifted, message, message_len);
    }
    memset(shifted + message_len, 0, n - message_len);
    remlen = poly_mod_be(shifted, n, g_be, glen, rem); /* remlen == n_check */
    /* codeword = message || (pad) || remainder. */
    if (message_len) {
        memcpy(out, message, message_len);
    }
    {
        size_t pad = n_check - remlen; /* 0 in the common case */
        memset(out + message_len, 0, pad);
        memcpy(out + message_len + pad, rem, remlen);
    }
    *out_len = n;
    return RS_OK;
}

/* ── decoding ─────────────────────────────────────────────────────────────── */
void rs_syndromes(const uint8_t *received, size_t received_len, size_t n_check,
                  uint8_t *out) {
    size_t i;
    for (i = 1; i <= n_check; i++) {
        out[i - 1] = poly_eval_be(received, received_len, gf256_power(2, (uint32_t)i));
    }
}

static int all_zero(const uint8_t *s, size_t n) {
    size_t i;
    for (i = 0; i < n; i++) {
        if (s[i] != 0) {
            return 0;
        }
    }
    return 1;
}

/* Berlekamp-Massey: fills `c_out` with the locator (LE, c[0]=1), sets *c_len,
 * returns the error count L. */
static size_t berlekamp_massey(const uint8_t *synds, size_t nsyn, uint8_t *c_out,
                               size_t *c_len) {
    uint8_t c[RS_BUF], b[RS_BUF];
    size_t clen = 1, blen = 1;
    size_t big_l = 0, x = 1;
    uint8_t b_scale = 1;
    size_t nn;
    c[0] = 1;
    b[0] = 1;
    for (nn = 0; nn < nsyn; nn++) {
        uint8_t d = synds[nn];
        size_t j, k;
        for (j = 1; j <= big_l; j++) {
            if (j < clen && nn >= j) {
                d = gf256_add(d, gf256_multiply(c[j], synds[nn - j]));
            }
        }
        if (d == 0) {
            x++;
        } else if (2 * big_l <= nn) {
            uint8_t tsave[RS_BUF];
            size_t tlen = clen;
            uint8_t scale = gf256_divide(d, b_scale);
            size_t shifted_len = x + blen;
            memcpy(tsave, c, clen);
            if (clen < shifted_len) {
                memset(c + clen, 0, shifted_len - clen);
                clen = shifted_len;
            }
            for (k = 0; k < blen; k++) {
                c[x + k] = gf256_add(c[x + k], gf256_multiply(scale, b[k]));
            }
            big_l = nn + 1 - big_l;
            memcpy(b, tsave, tlen);
            blen = tlen;
            b_scale = d;
            x = 1;
        } else {
            uint8_t scale = gf256_divide(d, b_scale);
            size_t shifted_len = x + blen;
            if (clen < shifted_len) {
                memset(c + clen, 0, shifted_len - clen);
                clen = shifted_len;
            }
            for (k = 0; k < blen; k++) {
                c[x + k] = gf256_add(c[x + k], gf256_multiply(scale, b[k]));
            }
            x++;
        }
    }
    memcpy(c_out, c, clen);
    *c_len = clen;
    return big_l;
}

size_t rs_error_locator(const uint8_t *syndromes, size_t nsyn, uint8_t *out) {
    size_t clen;
    /* A real syndrome sequence has length n_check <= 254; reject longer inputs
     * so the fixed Berlekamp-Massey buffers cannot overflow. */
    if (nsyn > 254) {
        return 0;
    }
    berlekamp_massey(syndromes, nsyn, out, &clen);
    return clen;
}

/* Inverse locator number for big-endian position p in a length-n codeword. */
static uint8_t inv_locator(size_t p, size_t n) {
    size_t exp = (p + 256 - n) % 255;
    return gf256_power(2, (uint32_t)exp);
}

/* Chien search: fills `positions` with error positions, returns the count. */
static size_t chien_search(const uint8_t *lambda, size_t llen, size_t n,
                           size_t *positions) {
    size_t p, count = 0;
    for (p = 0; p < n; p++) {
        uint8_t xi_inv = inv_locator(p, n);
        if (poly_eval_le(lambda, llen, xi_inv) == 0) {
            positions[count++] = p;
        }
    }
    return count;
}

/* Forney: error magnitudes for each position into `mags`. */
static rs_status forney(const uint8_t *lambda, size_t llen, const uint8_t *synds,
                        size_t nsyn, const size_t *positions, size_t npos,
                        size_t n, uint8_t *mags) {
    uint8_t omega_full[RS_BUF], lambda_prime[RS_BUF];
    size_t two_t = nsyn;
    size_t of_len = poly_mul_le(synds, nsyn, lambda, llen, omega_full);
    size_t omega_len = of_len < two_t ? of_len : two_t; /* Omega mod x^{2t} */
    size_t lplen = llen > 0 ? llen - 1 : 0;
    size_t j, i;
    memset(lambda_prime, 0, lplen ? lplen : 1);
    for (j = 0; j < llen; j++) {
        if (j % 2 == 1) {
            size_t out_idx = j - 1;
            if (out_idx < lplen) {
                lambda_prime[out_idx] = gf256_add(lambda_prime[out_idx], lambda[j]);
            }
        }
    }
    for (i = 0; i < npos; i++) {
        uint8_t xi_inv = inv_locator(positions[i], n);
        uint8_t omega_val = poly_eval_le(omega_full, omega_len, xi_inv);
        uint8_t lp_val = poly_eval_le(lambda_prime, lplen, xi_inv);
        if (lp_val == 0) {
            return RS_TOO_MANY_ERRORS;
        }
        mags[i] = gf256_divide(omega_val, lp_val);
    }
    return RS_OK;
}

rs_status rs_decode(const uint8_t *received, size_t received_len, size_t n_check,
                    uint8_t *out, size_t *out_len) {
    uint8_t synds[RS_BUF], lambda[RS_BUF], corrected[RS_BUF], mags[RS_BUF / 2];
    size_t positions[RS_BUF / 2];
    size_t t, n, k, llen, num_errors, npos, i;
    rs_status st;
    if (!valid_n_check(n_check)) {
        return RS_INVALID_INPUT;
    }
    /* A codeword cannot exceed the GF(256) block size; this also bounds every
     * fixed working buffer (received is copied into corrected[RS_BUF]). */
    if (received_len < n_check || received_len > 255) {
        return RS_INVALID_INPUT;
    }
    t = n_check / 2;
    n = received_len;
    k = n - n_check;

    rs_syndromes(received, received_len, n_check, synds);
    if (all_zero(synds, n_check)) {
        if (k) {
            memcpy(out, received, k);
        }
        *out_len = k;
        return RS_OK;
    }
    num_errors = berlekamp_massey(synds, n_check, lambda, &llen);
    if (num_errors > t) {
        return RS_TOO_MANY_ERRORS;
    }
    npos = chien_search(lambda, llen, n, positions);
    if (npos != num_errors) {
        return RS_TOO_MANY_ERRORS;
    }
    st = forney(lambda, llen, synds, n_check, positions, npos, n, mags);
    if (st != RS_OK) {
        return st;
    }
    memcpy(corrected, received, n);
    for (i = 0; i < npos; i++) {
        corrected[positions[i]] = gf256_add(corrected[positions[i]], mags[i]);
    }
    if (k) {
        memcpy(out, corrected, k);
    }
    *out_len = k;
    return RS_OK;
}
